### PR TITLE
41) realized authentication from DB

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 1]
```diff
@@ -1,40 +1,45 @@
...
 
     protected void configure(HttpSecurity http) throws Exception {
-        http.authorizeRequests()
-                .antMatchers("/register")
-                .permitAll()
-                .anyRequest()
-                .authenticated()
+        http
+                .authorizeRequests()
+                .and()
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java]: what are you authorizing here?

File: src/main/java/com/dev/cinema/security/CustomUserDetailsService.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Autowired
+    public CustomUserDetailsService(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        com.dev.cinema.model.User user = userService.findByEmail(email).get();
+        org.springframework.security.core.userdetails.User.UserBuilder builder =
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/security/CustomUserDetailsService.java:24]: move this to imports

**Review by Max-Pochepets — APPROVED** [review]

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 3]
```diff
@@ -1,40 +1,45 @@
...
+                .authorizeRequests()
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java]:
```suggestion
                .authorizeRequests()
```

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 4]
```diff
@@ -1,40 +1,45 @@
...
+                .antMatchers(HttpMethod.POST,"/movies/**").hasRole("ADMIN")
+                .antMatchers(HttpMethod.GET,"/movies/**").permitAll()
+                .anyRequest().authenticated()
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java]: let's add permitall for registration

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 5]
```diff
@@ -46,11 +45,9 @@ public User add(User user) {
     @Override
     public Optional<User> findByEmail(String email) {
         try (Session session = sessionFactory.openSession()) {
-            Query<User> userQuery = session
-                    .createQuery("from User where email =:email",
-                    User.class);
-            userQuery.setParameter("email", email);
-            return userQuery.uniqueResultOptional();
+            return session.createQuery("SELECT u FROM User u"
+                    + " LEFT JOIN FETCH u.roles WHERE u.email = :email", User.class)
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: inner join would better here, there should not be situation when user doesn't have role

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 6]
```diff
@@ -60,7 +57,9 @@ public User add(User user) {
     @Override
     public Optional<User> getById(Long id) {
         try (Session session = sessionFactory.openSession()) {
-            return Optional.ofNullable(session.get(User.class, id));
+            return session.createQuery("SELECT u FROM User u"
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java:61]: same here

File: src/main/java/com/dev/cinema/security/CustomUserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    private final UserService userService;
+
+    @Autowired
+    public CustomUserDetailsService(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        com.dev.cinema.model.User user = userService.findByEmail(email).get();
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/security/CustomUserDetailsService.java]: let's throw UsernameNotFoundException instead of NoSuchElementException

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java (hunk 5 above)
Comment (RomanDubovskyi) [reply src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: you didnt fix this comment - change LEFT JOIN to JOIN and don't remove .setParameter()

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 8]
```diff
@@ -58,7 +58,7 @@ public User add(User user) {
     public Optional<User> getById(Long id) {
         try (Session session = sessionFactory.openSession()) {
             return session.createQuery("SELECT u FROM User u"
-                    + " LEFT JOIN FETCH u.roles WHERE u.id = :id", User.class)
+                    + " WHERE u.id = :id", User.class)
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java:63]: same here

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/security/CustomUserDetailsService.java (hunk 7 above)
Comment (RomanDubovskyi) [reply src/main/java/com/dev/cinema/security/CustomUserDetailsService.java]: this comment wasn't fixed

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 9]
```diff
@@ -1,26 +1,38 @@
...
 
     protected void configure(HttpSecurity http) throws Exception {
-        http.authorizeRequests()
+        http
+                .authorizeRequests()
                 .antMatchers("/register")
-                .permitAll()
+                .authenticated()
+                .antMatchers(HttpMethod.POST,"/movies/**").hasRole("ADMIN")
+                .antMatchers(HttpMethod.GET,"/movies/**").permitAll()
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java:35]: fix permissions according to the message that Sofia has written in general channel, rest seems to be good

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**